src/smt/sygus_solver.cpp

```diff
@@ -185,45 +192,76 @@ Result SygusSolver::checkSynth(Assertions& as)
         "Cannot make check-synth commands when incremental solving is enabled");
   }
   Trace("smt") << "SygusSolver::checkSynth" << std::endl;
-  std::vector<Node> query;
+  // if applicable, check if the subsolver is the correct one
+  if (usingSygusSubsolver() && d_subsolverCd.get() != d_subsolver.get())
+  {
+    // this can occur if we backtrack to a place where we were using a different
+    // subsolver than the current one. In this case, we should reconstruct
+    // the subsolver.
+    d_sygusConjectureStale = true;
+  }
   if (d_sygusConjectureStale)
   {
     NodeManager* nm = NodeManager::currentNM();
     // build synthesis conjecture from asserted constraints and declared
     // variables/functions
     Trace("smt") << "Sygus : Constructing sygus constraint...\n";
-    Node body = nm->mkAnd(d_sygusConstraints);
+    Node body = nm->mkAnd(listToVector(d_sygusConstraints));
     // note that if there are no constraints, then assumptions are irrelevant
     if (!d_sygusConstraints.empty() && !d_sygusAssumps.empty())
     {
-      Node bodyAssump = nm->mkAnd(d_sygusAssumps);
+      Node bodyAssump = nm->mkAnd(listToVector(d_sygusAssumps));
       body = nm->mkNode(IMPLIES, bodyAssump, body);
     }
     body = body.notNode();
     Trace("smt") << "...constructed sygus constraint " << body << std::endl;
     if (!d_sygusVars.empty())
     {
-      Node boundVars = nm->mkNode(BOUND_VAR_LIST, d_sygusVars);
+      Node boundVars = nm->mkNode(BOUND_VAR_LIST, listToVector(d_sygusVars));
       body = nm->mkNode(EXISTS, boundVars, body);
       Trace("smt") << "...constructed exists " << body << std::endl;
     }
     if (!d_sygusFunSymbols.empty())
     {
-      body =
-          quantifiers::SygusUtils::mkSygusConjecture(d_sygusFunSymbols, body);
+      body = quantifiers::SygusUtils::mkSygusConjecture(
+          listToVector(d_sygusFunSymbols), body);
     }
     Trace("smt") << "...constructed forall " << body << std::endl;
 
     Trace("smt") << "Check synthesis conjecture: " << body << std::endl;
 
     d_sygusConjectureStale = false;
 
-    // TODO (project #7): if incremental, we should push a context and assert
-    query.push_back(body);
-  }
+    d_conj = body;
+
+    // if we are using a subsolver, initialize it now
+    if (usingSygusSubsolver())
+    {
+      // we generate a new solver engine to do the SyGuS query
+      initializeSygusSubsolver(d_subsolver, as);
 
-  Result r = d_smtSolver.checkSatisfiability(as, query, false);
+      // store the pointer (context-dependent)
+      d_subsolverCd = d_subsolver.get();
 
+      // also assert the internal SyGuS conjecture
+      d_subsolver->assertFormula(d_conj);
+    }
+  }
+  else
+  {
+    Assert(d_subsolver != nullptr);
+  }
```

Comment:
You could delete this and add `Assert(d_sygusConjectureStale || d_subsolver != nullptr);` before the `if`

Reply:
There will be code added to this block (e.g. to block the current solution).